main.js

```diff
@@ -25,31 +25,74 @@ const artifact = require('@actions/artifact');
       charmcraft_channel,
     ]);
 
-    await exec.exec('charmcraft', ['pack', '--destructive-mode', '-q']);
+    chdir(charm_path);
+    core.exportVariable('CHARMCRAFT_AUTH', credentials);
+
+    const metadata = yaml.load(fs.readFileSync('metadata.yaml'));
+
+    const name = metadata.name;
+    const images = Object.entries(metadata.resources)
+      .filter(([_, res]) => res.type === 'oci-image')
+      .map(([name, res]) => [name, res['upstream-source']]);
+
+    const branch = (
+      await exec.getExecOutput('git', ['rev-parse', '--abbrev-ref', 'HEAD'])
+    ).stdout.trim();
+
+    let channel;
+
+    if (['master', 'main'].includes(branch)) {
+      channel = 'latest/edge';
+    } else if (branch.startsWith('track/')) {
+      channel = branch.replace('track/', '') + '/edge';
+    } else if (branch.startsWith('branch/')) {
+      channel = 'latest/edge/' + branch.replace('branch/', '');
+    } else {
+      core.setFailed(
+        `Invalid branch name ${branch}. ` +
+          'The branch name must be master, main, or begin with ' +
+          "'track/' or 'branch/' for this action to publish a charm."
+      );
+      return;
+    }
+
+    const revisions = await Promise.all(
+      images.map(async ([resource_name, resource_image]) => {
+        await exec.exec('docker', ['pull', resource_image]);
```

Comment:
Should we move the `docker pull`/`charmcraft upload-resource` commands to happen near the end rather than the beginning?  If all else is equal, I think we want the commands in order of `most likely to fail --> least likely to fail` so when someone is debugging they can spot problems quickly, and the pulling/uploading feels pretty robust compared to the packing